policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+        initNamespaces = _get_api_list("namespaces")
+        initServices = _get_api_list("services")
+        initPods = _get_api_list("pods")
+        initEndpoints = _get_api_list("endpoints")
+
+        for ns in initNamespaces["items"]:
+            self.process_resource(action=CMD_ADDED,
+                                  kind=KIND_NAMESPACE,
+                                  target=ns)
+
+        for svc in initServices["items"]:
+            self.process_resource(action=CMD_ADDED,
+                                  kind=KIND_SERVICE,
+                                  target=svc)
+
+        for pod in initPods["items"]:
+            self.process_resource(action=CMD_ADDED,
+                                  kind=KIND_POD,
+                                  target=pod)
+
+        for ep in initEndpoints["items"]:
+            self.process_resource(action=CMD_ADDED,
+                                  kind=KIND_ENDPOINTS,
+                                  target=ep)
+
+        self.nsResourceVersion = initEndpoints["metadata"]["resourceVersion"]
+        self.svcResourceVersion = initServices["metadata"]["resourceVersion"]
+        self.poResourceVersion = initPods["metadata"]["resourceVersion"]
+        self.epResourceVersion = initNamespaces["metadata"]["resourceVersion"]
+
+        _log.info("Service Resource Version = %s" % self.svcResourceVersion)
+        _log.info("Namespace Resource Version = %s" % self.nsResourceVersion)
+        _log.info("Pod Resource Version = %s" % self.poResourceVersion)
+        _log.info("Endpoints Resource Version = %s" % self.epResourceVersion)
+
+    def init_threads(self):
+        # Assert these have not been initialized yet.
+        assert not self.namespace_watcher
+        assert not self.service_watcher
+        assert not self.pod_watcher
+        assert not self.endpoints_watcher
+
+        self.namespace_watcher = Thread(target=_keep_watch,
+                                        args=(self.watcher_queue, "namespaces", self.nsResourceVersion))
+        self.service_watcher = Thread(target=_keep_watch,
+                                      args=(self.watcher_queue, "services", self.svcResourceVersion))
+        self.pod_watcher = Thread(target=_keep_watch,
+                                  args=(self.watcher_queue, "pods", self.poResourceVersion))
+        self.endpoints_watcher = Thread(target=_keep_watch,
+                                        args=(self.watcher_queue, "endpoints", self.epResourceVersion))
+
+        self.namespace_watcher.daemon = True
+        self.service_watcher.daemon = True
+        self.pod_watcher.daemon = True
+        self.endpoints_watcher.daemon = True
+
+    def read_updates(self):
+        """
+        Pulls an update off the Queue and processes it.
+        If no responses remain in the queue, it will trigger a resync for all resource lists.
+        """
+        try:
+            response = self.watcher_queue.get_nowait()
```

Comment:
If there are no responses in the queue for a while, I think this will tight-loop and keep attempting to resync.

I think we can just use self.watcher_queue.get() so that we wait for the queue to be populated.  This will prevent tight-looping when the queue is empty.